R/filterReadsBam.R

```diff
@@ -0,0 +1,174 @@
+#'     fraction of low-confidence modified base calls in a read. Reads with
+#'     a fraction of low confidence calls greater than this value will be
+#'     filtered out.
+#' @param maxEntropy A numeric scalar representing the largest acceptable
+#'     read-level entropy. Reads with entropy above this value will be filtered
+#'     out. A value of \code{Inf} deactivates the entropy filter.
+#' @param BPPARAM A \code{\link[BiocParallel]{BiocParallelParam}} object that
+#'     controls the number of parallel CPU threads to use for some of the steps
+#'     in \code{filterReadsBam()}. The default value is
+#'     (\code{\link[BiocParallel]{MulticoreParam}(4L, RNGseed = 42L)}).
+#' @param verbose Logical scalar. If \code{TRUE}, report on progress.
+#'
+#' @return A \code{data.frame} with one row per \code{infile} giving the numbers
+#'     of filtered out records per reason for exclusion.
+#'
+#' @examples
+#' modbamfiles <- system.file("extdata", c("6mA_1_10reads.bam", "6mA_2_10reads.bam"),
+#'                            package = "footprintR")
+#' filtbamfiles <- tempfile(fileext = rep(".bam", length(modbamfiles)))
+#' res <- filterReadsBam(infiles = modbamfiles, outfiles = filtbamfiles,
+#'                       modbase = "a", indexOutfiles = FALSE, minReadLength = 6746,
+#'                       minAlignedLength = 6896, minAlignedFraction = 0.56,
+#'                       minQscore = 9.7, maxFracLowConf = 0.11, maxEntropy = 0.29,
+#'                       BPPARAM = BiocParallel::SerialParam(), verbose = TRUE)
+#' res
+#' unlink(filtbamfiles)
+#'
+#' @author Michael Stadler
+#'
+#' @importFrom BiocParallel MulticoreParam bplapply bpnworkers bpworkers<-
+#' @importFrom cli cli_abort cli_alert_info
+#'
+#' @export
+filterReadsBam <- function(infiles,
+                           outfiles,
+                           modbase,
+                           indexOutfiles = TRUE,
+                           minReadLength = 0,
+                           minAlignedLength = 0,
+                           minAlignedFraction = 0,
+                           minQscore = 0.0,
+                           maxFracLowConf = 1.0,
+                           maxEntropy = Inf,
+                           BPPARAM = MulticoreParam(4L, RNGseed = 42L),
+                           verbose = FALSE) {
+    # validate arguments
+    .assertVector(x = infiles, type = "character")
+    if (any(i <- !file.exists(infiles))) {
+        cli_abort(paste0("not all `infiles` exist: ",
+                         paste(infiles[i], collapse = ", ")))
+    }
+    if (is.null(names(infiles))) {
+        names(infiles) <- paste0("s", seq_along(infiles))
+    } else if (any(duplicated(names(infiles)))) {
+        cli_abort("`names(infiles)` are not unique")
+    }
+    .assertVector(x = outfiles, type = "character", len = length(infiles))
+    if (any(i <- file.exists(outfiles))) {
+        cli_abort(paste0("existing `outfiles` would be overwritten: ",
```

Comment:
Should we have an option to overwrite existing outfiles?

Reply:
Good idea. I have added an `overwriteOutfiles = FALSE` argument in db613e59